src/typechecker.erl

```diff
@@ -3438,16 +3438,76 @@ refinable(_) ->
+check_guard_call(P, is_list, [{var, _, Var}]) -> #{Var => {type, P, list, []}};
+check_guard_call(P, is_map, [{var, _, Var}]) -> #{Var => {type, P, map, []}};
+check_guard_call(P, is_number, [{var, _, Var}]) -> #{Var => {type, P, number, []}};
+check_guard_call(P, is_pid, [{var, _, Var}]) -> #{Var => {type, P, pid, []}};
+check_guard_call(P, is_port, [{var, _, Var}]) -> #{Var => {type, P, port, []}};
+check_guard_call(P, is_record, [{var, _, Var}, {atom, _, Record}]) ->
+    #{Var => {type, P, record, [{atom, P, Record}]}};
+check_guard_call(P, is_record, [{var, _, Var}, {atom, _, Record}, _]) ->
+    #{Var => {type, P, record, [{atom, P, Record}]}};
+check_guard_call(P, is_reference, [{var, _, Var}]) -> #{Var => {type, P, reference, []}};
+check_guard_call(P, is_tuple, [{var, _, Var}]) -> #{Var => {type, P, tuple, []}};
+check_guard_call(_P, _Fun, _Vars) ->
+    #{}.
+
+-spec check_guard(#env{}, term()) -> map().
+check_guard(_Env, {call, P, {atom, _, Fun}, Vars}) ->
+    check_guard_call(P, Fun, Vars);
+check_guard(_Env, {call, P, {remote,_,_,{atom, _, Fun}}, Vars}) ->
+    check_guard_call(P, Fun, Vars);
+check_guard(Env, {op, _OrElseAnno, 'orelse', Call1, Call2}) ->
+    G1 = check_guard(Env, Call1),
+    G2 = check_guard(Env, Call2),
+    Lub = fun(_K, Ty1, Ty2) ->
+        UTy = {type, erl_anno:new(0), union, [Ty1, Ty2]},
+        NTy = normalize(UTy, Env#env.tenv),
+        NTy
+    end,
+    union_var_binds_help([G1, G2], Lub);
+check_guard(Env, {op, _AndAlsoAnno, 'andalso', Call1, Call2}) ->
+    G1 = check_guard(Env, Call1),
+    G2 = check_guard(Env, Call2),
+    union_var_binds([G1, G2], Env#env.tenv);
+check_guard(Env, Guard) ->
+    {_Ty, VB, _Cs} = type_check_expr(Env, Guard), % Do we need to thread the Env?
+    VB.
+
+%% The different guards use glb
+-spec check_guards_sequence(#env{}, list()) -> map().
+check_guards_sequence(Env, GuardSeq) ->
+    RefTys = union_var_binds(
+        lists:map(fun (Guard) ->
+            check_guard(Env, Guard)
+                  end, GuardSeq),
+        Env#env.tenv),
+    maps:merge(Env#env.venv, RefTys).
+
 %% TODO: implement proper checking of guards.
+-spec check_guards(#env{}, list()) -> map().
+check_guards(Env, []) -> #{};
 check_guards(Env, Guards) ->
-    union_var_binds(
-      lists:map(fun (GuardSeq) ->
-                        union_var_binds(
-                          lists:map(fun (Guard) ->
-                                                {_Ty, VB, _Cs} = type_check_expr(Env, Guard), % Do we need to thread the Env?
-                                                VB
-                                    end, GuardSeq), Env#env.tenv)
-                end, Guards), Env#env.tenv).
+    Lub = fun(_K, Ty1, Ty2) ->
+        UTy = {type, erl_anno:new(0), union, [Ty1, Ty2]},
+        NTy = normalize(UTy, Env#env.tenv),
+        NTy
+    end,
+    Tys = lists:map(fun (GuardSeq) ->
+        check_guards_sequence(Env, GuardSeq)
+    end, Guards),
+    Ty = union_var_binds_help(Tys, Lub),
```

Comment:
This line and the definition of Lub is the same in the code for `orelse`, so I think the common code could be reused, perhaps wrapped in a function named something like `union_var_binds_using_lub`?

Also, the variable name `Ty` and `Tys` here are confusing, because these are not types. They are var binds (maps) and list of such.

----

I'm not convinced though that `union_var_binds_help(Tys, Lub)` correctly merges `X :: integer()` with `Y :: integer()` into the empty set of var binds. (`union_var_binds_help` calls `gradualizer_lib:merge_with/3` which is using `maps:update_with/4`, which if I'm reading correctly would return `X :: integer(), Y :: integer()` in this case.)

Thus, I think `union_var_binds_using_lub` (or with a better name) needs to be implemented differently, so that only variables which are defined in all var bind maps are merged using Lub. Otherwise, they should be omitted.

To verify this, I'd like to see a *should fail* test case like this:

```Erlang
-spec f(integer() | atom(), integer() | atom()) -> integer().
f(X, Y) when is_integer(X); is_integer(Y) -> X + Y;
f(_, _) -> 42.
```

A similar guard which should prevent refinement is P in the following:

```Erlang
-spec g(integer() | atom(), boolean()) -> integer().
g(X, P) when is_integer(X); P -> X.
```

Reply:
Fixed the first part, thanks for the feedback. Looking into the second part now.

Reply:
To make things clear, `;` is `orelse`.  I will add test cases for both to make sure everything works.

Reply:
Thanks for that man, good finding. Managed to make sure both `orelse` and `andalso` correctly bind stuff. Check the new test cases. Resolving this.